File: Algorithms/946_Validate_Stack_Sequences/Solution.scala
import scala.collection.mutable

object Solution {
  def validateStackSequences(pushed: Array[Int], popped: Array[Int]): Boolean = {
    val stk = mutable.Stack[Int]()

    var i = 0
    for (element <- pushed) {
      stk.push(element)
      while (stk.nonEmpty && stk.top == popped(i)) {
        stk.pop()
        i += 1
      }
    }

    stk.isEmpty
  }
}

File: Algorithms/946_Validate_Stack_Sequences/solution.py
from typing import *


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stk: List[int] = []
        i = 0
        for element in pushed:
            stk.append(element)
            while stk and stk[-1] == popped[i]:
                stk.pop()
                i += 1
        return not stk
